Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_notifyPropose/NotifyProposeResponder.swift

```diff
@@ -0,0 +1,97 @@
+
+import Foundation
+import Combine
+
+class NotifyProposeResponder {
+    enum Errors: Error {
+        case recordForIdNotFound
+        case malformedRequestParams
+        case subscriptionNotFound
+    }
+    private let networkingInteractor: NetworkInteracting
+    private let kms: KeyManagementServiceProtocol
+    private let logger: ConsoleLogging
+    private let pushSubscribeRequester: PushSubscribeRequester
+    private let rpcHistory: RPCHistory
+    private var subscriptionResponsePublisher: AnyPublisher<Result<PushSubscription, Error>, Never>
+
+    private var publishers = [AnyCancellable]()
+
+    init(networkingInteractor: NetworkInteracting,
+         kms: KeyManagementServiceProtocol,
+         logger: ConsoleLogging,
+         pushSubscribeRequester: PushSubscribeRequester,
+         rpcHistory: RPCHistory,
+        pushSubscribeResponseSubscriber: PushSubscribeResponseSubscriber
+    ) {
+        self.networkingInteractor = networkingInteractor
+        self.kms = kms
+        self.logger = logger
+        self.pushSubscribeRequester = pushSubscribeRequester
+        self.subscriptionResponsePublisher = pushSubscribeResponseSubscriber.subscriptionPublisher
+        self.rpcHistory = rpcHistory
+    }
+
+    func approve(requestId: RPCID, onSign: @escaping SigningCallback) async throws {
+
+        logger.debug("NotifyProposeResponder: approving proposal")
+
+        guard let requestRecord = rpcHistory.get(recordId: requestId) else { throw Errors.recordForIdNotFound }
+        let proposal = try requestRecord.request.params!.get(NotifyProposeParams.self)
+
+        let subscriptionAuthWrapper = try await pushSubscribeRequester.subscribe(metadata: proposal.metadata, account: proposal.account, onSign: onSign)
+
+        var pushSubscription: PushSubscription!
+        try await withCheckedThrowingContinuation { continuation in
```

Comment:
As we do not have subscription in this point, may be we should listen subscriptionResponsePublisher separately? Not inside `approve` method

Reply:
🤔 hmm, like how? I kind of want to block a thread here and wait for only the first subscription